File: WaterClockApp/app/src/main/java/br/com/waterclockapp/ui/camera/CameraFragment.java
package br.com.waterclockapp.ui.camera;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.ImageFormat;
import android.graphics.SurfaceTexture;
import android.hardware.camera2.CameraCaptureSession;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CameraManager;
import android.hardware.camera2.CameraMetadata;
import android.hardware.camera2.CaptureRequest;
import android.hardware.camera2.TotalCaptureResult;
import android.hardware.camera2.params.StreamConfigurationMap;
import android.media.Image;
import android.media.ImageReader;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.HandlerThread;
import android.provider.MediaStore;
import android.util.Size;
import android.util.SparseIntArray;
import android.view.LayoutInflater;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;

import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

import br.com.waterclockapp.R;
import br.com.waterclockapp.ui.home.HomeContract;
import br.com.waterclockapp.util.ConstantsKt;

public class CameraFragment extends Fragment {

    private static final int GALLERY_REQUEST_CODE = 756;
    private Size previewSize;
    private Size jpegSizes[] = null;

    private static final int REQUEST_CODE = 200;
    private static final String CAMERA_PERMISSION = Manifest.permission.CAMERA;
    private static final String CAPTURE_IMAGE_PERMISSION = Manifest.permission.WRITE_EXTERNAL_STORAGE;

    private HomeContract.View viewHome;

    private TextureView textureView;
    private CameraDevice cameraDevice;
    private CaptureRequest.Builder previewBuilder;
    private CameraCaptureSession previewSession;
    private Button getPicture;
    private Button getGallery;

    private static final SparseIntArray ORIENTATIONS = new SparseIntArray();

    static {
        ORIENTATIONS.append(Surface.ROTATION_0,90);
        ORIENTATIONS.append(Surface.ROTATION_90,0);
        ORIENTATIONS.append(Surface.ROTATION_180,270);
        ORIENTATIONS.append(Surface.ROTATION_270,180);
    }

    public CameraFragment(HomeContract.View viewHome) {
        this.viewHome = viewHome;
    }


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        textureView = (TextureView) view.findViewById(R.id.surfaceViewCamera);
        textureView.setSurfaceTextureListener(surfaceTextureListener);

        getPicture = view.findViewById(R.id.buttonSavaCamera);
        getPicture.setOnClickListener(v -> {
            getPicture();
        });

        getGallery = view.findViewById(R.id.buttonOpenGallery);
        getGallery.setOnClickListener( v -> getGallery());
    }

    private void getGallery() {
        Intent intent = new Intent(Intent.ACTION_PICK);
        intent.setType("image/*");
        String[] mineTypes = {"image/jpeg", "image/png"};
        intent.putExtra(Intent.EXTRA_MIME_TYPES, mineTypes);
        startActivityForResult(intent, GALLERY_REQUEST_CODE);
        //openCamera();

    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if(resultCode == Activity.RESULT_OK){
            switch (requestCode){
                case GALLERY_REQUEST_CODE:
                    Uri selectedImage = data.getData();
                    //
                    String[] filePathColumn = {MediaStore.Images.Media.DATA};
                    Cursor cursor = getActivity().getContentResolver().query(selectedImage, filePathColumn, null, null, null);
                    cursor.moveToFirst();
                    int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
                    String imgDecodableString = cursor.getString(columnIndex);
                    cursor.close();
                    Bitmap bitmap = BitmapFactory.decodeFile(imgDecodableString);
                    showDialogSend(bitmap);
                    Toast.makeText(getActivity(), "PEGOU FOTO " + bitmap.toString(), Toast.LENGTH_LONG).show();
                    break;
            }
        }
        openCamera();

    }

    private void getPicture() {
        if(cameraDevice == null) return;
        CameraManager manager = (CameraManager) getActivity().getSystemService(Context.CAMERA_SERVICE);
        try{
            CameraCharacteristics characteristics = manager.getCameraCharacteristics(cameraDevice.getId());
            if(characteristics != null){
                jpegSizes = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)
                        .getOutputSizes(ImageFormat.JPEG);
            }
            int width = 640, height = 480;
            if(jpegSizes != null && jpegSizes.length > 0){
                width = jpegSizes[0].getWidth();
                height = jpegSizes[0].getHeight();
            }
            ImageReader reader = ImageReader.newInstance(width, height, ImageFormat.JPEG, 1);
            List<Surface> outputSurfaces = new ArrayList<>(2);
            outputSurfaces.add(reader.getSurface());
            outputSurfaces.add(new Surface(textureView.getSurfaceTexture()));

            final  CaptureRequest.Builder captureBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_STILL_CAPTURE);
            captureBuilder.addTarget(reader.getSurface());
            captureBuilder.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO);

            int rotation = getActivity().getWindowManager().getDefaultDisplay().getRotation();
            captureBuilder.set(CaptureRequest.JPEG_ORIENTATION, ORIENTATIONS.get(rotation));

            ImageReader.OnImageAvailableListener imageAvailableListener=new ImageReader.OnImageAvailableListener() {
                @Override
                public void onImageAvailable(ImageReader reader) {
                    Image image = null;
                    try {
                        image = reader.acquireLatestImage();
                        ByteBuffer buffer = image.getPlanes()[0].getBuffer();
                        byte[] bytes = new byte[buffer.capacity()];
                        buffer.get(bytes);
                        save(bytes);
                    } catch (Exception ee) {
                    }
                    finally {
                        if(image!=null)
                            image.close();
                    }
                }
                void save(byte[] bytes)
                {
                    File file12=getOutputMediaFile();
                    OutputStream outputStream=null;
                    try
                    {
                        outputStream= new FileOutputStream(file12);
                        outputStream.write(bytes);

                        Toast.makeText(getActivity(), "PEGOU FOTO", Toast.LENGTH_LONG).show();
                        Bitmap image = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
                        showDialogSend(image);

                    }catch (Exception e)
                    {
                        e.printStackTrace();
                    }finally {
                        try {
                            if (outputStream != null)
                                outputStream.close();
                        }catch (Exception e){}
                    }
                }
            };

            HandlerThread handlerThread=new HandlerThread("takepicture");
            handlerThread.start();
            final Handler handler=new Handler(handlerThread.getLooper());
            reader.setOnImageAvailableListener(imageAvailableListener,handler);
            final CameraCaptureSession.CaptureCallback  previewSSession=new CameraCaptureSession.CaptureCallback() {
                @Override
                public void onCaptureStarted(CameraCaptureSession session, CaptureRequest request, long timestamp, long frameNumber) {
                    super.onCaptureStarted(session, request, timestamp, frameNumber);
                }
                @Override
                public void onCaptureCompleted(CameraCaptureSession session, CaptureRequest request, TotalCaptureResult result) {
                    super.onCaptureCompleted(session, request, result);
                    startCamera();
                }
            };
            cameraDevice.createCaptureSession(outputSurfaces, new CameraCaptureSession.StateCallback() {
                @Override
                public void onConfigured(CameraCaptureSession session) {
                    try
                    {
                        session.capture(captureBuilder.build(),previewSSession,handler);
                    }catch (Exception e)
                    {

                    }
                }
                @Override
                public void onConfigureFailed(CameraCaptureSession session) {
                }
            },handler);
        }catch (Exception e){
            e.printStackTrace();
        }
    }



    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_camera, container, false);
    }



    //OPEN CAMERA
    public void openCamera(){

        if((ContextCompat.checkSelfPermission(getActivity(), CAMERA_PERMISSION) != PackageManager.PERMISSION_GRANTED)
        || (ContextCompat.checkSelfPermission(getActivity(), CAPTURE_IMAGE_PERMISSION) != PackageManager.PERMISSION_GRANTED)) requestCameraPermission();
        CameraManager manager = (CameraManager) getActivity().getSystemService(Context.CAMERA_SERVICE);
        try {
            String cameraId = manager.getCameraIdList()[0];
            CameraCharacteristics characteristics = manager.getCameraCharacteristics(cameraId);
            StreamConfigurationMap map = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);
            previewSize = (map != null ? map.getOutputSizes(SurfaceTexture.class) : new Size[0])[0];
            manager.openCamera(cameraId, stateCallback, null);
        }catch (Exception e){
            e.printStackTrace();
        }
    }

    private void requestCameraPermission() {
        ActivityCompat.requestPermissions(getActivity(), new String[]{CAMERA_PERMISSION,
                CAPTURE_IMAGE_PERMISSION}, REQUEST_CODE);
    }

    //SURFACE TEXTURE LISTENER
    private TextureView.SurfaceTextureListener surfaceTextureListener = new TextureView.SurfaceTextureListener() {
        @Override
        public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
            openCamera();
        }

        @Override
        public void onSurfaceTextureSizeChanged(SurfaceTexture surface, int width, int height) {

        }

        @Override
        public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
            return false;
        }

        @Override
        public void onSurfaceTextureUpdated(SurfaceTexture surface) {

        }
    };

    private CameraDevice.StateCallback stateCallback = new CameraDevice.StateCallback() {
        @Override
        public void onOpened(@NonNull CameraDevice camera) {
            cameraDevice = camera;
            startCamera();
        }

        @Override
        public void onDisconnected(@NonNull CameraDevice camera) {

        }

        @Override
        public void onError(@NonNull CameraDevice camera, int error) {

        }
    };

    @Override
    public void onPause() {
        super.onPause();
        if (cameraDevice != null) cameraDevice.close();

    }

    void startCamera(){
        if(cameraDevice == null || !textureView.isAvailable() || previewSize == null) return;

        SurfaceTexture texture = textureView.getSurfaceTexture();
        if(texture == null) return;
        texture.setDefaultBufferSize(previewSize.getWidth(), previewSize.getHeight());
        Surface surface = new Surface(texture);
        try{
            previewBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);
            previewBuilder.addTarget(surface);
            cameraDevice.createCaptureSession(Arrays.asList(surface), new CameraCaptureSession.StateCallback() {
                @Override
                public void onConfigured(@NonNull CameraCaptureSession session) {
                    previewSession = session;
                    try{
                        getChangedPreview();
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }

                @Override
                public void onConfigureFailed(@NonNull CameraCaptureSession session) {

                }
            }, null);
        }catch (Exception e){
            e.printStackTrace();
        }

    }

    void getChangedPreview() throws Exception{
        if(cameraDevice == null) return;
        previewBuilder.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO);
        HandlerThread thread = new HandlerThread("changed preview");
        thread.start();
        Handler handler = new Handler(thread.getLooper());
        previewSession.setRepeatingRequest(previewBuilder.build(), null, handler);
    }

    private static File getOutputMediaFile(){
        File mediaStorageDir = new File(Environment.getExternalStorageDirectory(), "WaterClock");
        if(!mediaStorageDir.exists()){
            if(!mediaStorageDir.mkdirs()) return null;
        }

        @SuppressLint("SimpleDateFormat") String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
        File mediaFile;
        mediaFile = new File(mediaStorageDir.getPath() + File.separator + "WC_"+timeStamp+".jpg");

        System.out.println("SAVE WITH SUCCESS");
        return mediaFile;
    }

    private void showDialogSend(Bitmap image){
        Dialog dialog = new Dialog(getContext(), R.style.CustomAlertDialog);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        dialog.setContentView(R.layout.send_image_dialog);
        dialog.setCancelable(false);
        ((ImageView) dialog.findViewById(R.id.imageViewSendPhoto)).setImageBitmap(image);
        ((TextView) dialog.findViewById(R.id.textViewDialogQuestion)).setText(ConstantsKt.DIALOG_SEND_IMAGE);
        dialog.show();

        dialog.findViewById(R.id.buttonNoChoose).setOnClickListener(v-> dialog.dismiss());
        dialog.findViewById(R.id.buttonYesChoose).setOnClickListener(v-> {
            Toast.makeText(getContext(), "ENVIADO", Toast.LENGTH_LONG).show();
            dialog.dismiss();
        } );
    }


    public void enableNavigation(Boolean key){
        viewHome.enableNavigation(key);
    }

}

File: WaterClockApp/app/src/main/java/br/com/waterclockapp/ui/camera/CameraFragment.kt
package br.com.waterclockapp.ui.camera

import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Camera
import android.graphics.ImageFormat
import android.graphics.SurfaceTexture
import android.hardware.camera2.*
import android.hardware.camera2.params.StreamConfigurationMap
import android.media.Image
import android.media.ImageReader
import android.os.Bundle
import android.os.Environment
import android.os.Handler
import android.os.HandlerThread
import android.system.Os.close
import android.util.Size
import android.util.SparseIntArray
import android.view.*
import android.widget.Button
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import br.com.waterclockapp.R
import br.com.waterclockapp.ui.HomeActivity
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream
import java.nio.ByteBuffer

class CameraFragment : Fragment() {

    private val REQUESTCODECAMERA: Int = 101
    val ORIENTATIONS : SparseIntArray  = SparseIntArray().apply{
        append(Surface.ROTATION_0, 90)
        append(Surface.ROTATION_90, 0)
        append(Surface.ROTATION_180, 270)
        append(Surface.ROTATION_270, 180)

    }


    private lateinit var cameraId: String
    var cameraDevice:CameraDevice? = null
    lateinit var cameraCaptureSession: CameraCaptureSession
    lateinit var captureRequest: CaptureRequest
    lateinit var captureRequestBuilder:CaptureRequest.Builder
    private lateinit var size: Size
    private lateinit var imageReader: ImageReader
    private lateinit var file: File
    var backgroundHandler: Handler? = null
    var backgroundThread: HandlerThread? = null

    private lateinit var textureViewCamera : TextureView
    private lateinit var buttonSavaCamera : Button


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_camera, container, false)
        textureViewCamera = view.findViewById(R.id.textureViewCamera)
        buttonSavaCamera = view.findViewById(R.id.buttonSavaCamera)
        return inflater.inflate(R.layout.fragment_camera, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        textureViewCamera.surfaceTextureListener = textureListener
        buttonSavaCamera.setOnClickListener {
            takePicture()
        }
    }

    private val textureListener : TextureView.SurfaceTextureListener = object : TextureView.SurfaceTextureListener {
        override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture?, width: Int, height: Int) {

        }

        override fun onSurfaceTextureUpdated(surface: SurfaceTexture?) {

        }

        override fun onSurfaceTextureDestroyed(surface: SurfaceTexture?): Boolean {
            return false
        }

        override fun onSurfaceTextureAvailable(surface: SurfaceTexture?, width: Int, height: Int) {
           try {
               openCamera()
           }catch (e:CameraAccessException){

           }
        }
    }

    private val stateCallback: CameraDevice.StateCallback = object : CameraDevice.StateCallback(){
        override fun onOpened(camera: CameraDevice) {
            cameraDevice = camera
            createCameraPreview()
        }

        override fun onDisconnected(camera: CameraDevice) {
            cameraDevice?.close()
        }

        override fun onError(camera: CameraDevice, error: Int) {
            cameraDevice?.close()
            cameraDevice = null
        }

    }

    private fun createCameraPreview() {
        val texture : SurfaceTexture = textureViewCamera.surfaceTexture
        texture.setDefaultBufferSize(size.width, size.height)
        val surface: Surface = Surface(texture)

        captureRequestBuilder = (cameraDevice?.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW) ?: captureRequestBuilder.addTarget(surface)) as CaptureRequest.Builder

        cameraDevice?.createCaptureSession(listOf(surface), object : CameraCaptureSession.StateCallback(){
            override fun onConfigureFailed(session: CameraCaptureSession) {
                if(cameraDevice == null)
                    return

                cameraCaptureSession = session
                updatePreview()
            }

            override fun onConfigured(session: CameraCaptureSession) {
                Toast.makeText(activity, "Configuration Changed", Toast.LENGTH_LONG).show()
            }



        }, null)

    }

    private fun updatePreview() {
        if(cameraDevice == null) return

        captureRequestBuilder.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO)
        cameraCaptureSession.setRepeatingRequest(captureRequestBuilder.build(), null, backgroundHandler)

    }

    private fun openCamera() {

        val cameraActivity = activity
        if(cameraActivity == null || cameraActivity.isFinishing) return
        val cameraManager: CameraManager = cameraActivity.getSystemService(Context.CAMERA_SERVICE) as CameraManager
        cameraId = cameraManager.cameraIdList[0]
        val characteristics: CameraCharacteristics = cameraManager.getCameraCharacteristics(cameraId)
        val map : StreamConfigurationMap = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)
        size = map.getOutputSizes(SurfaceTexture::class.java)[0]
        if(ActivityCompat.checkSelfPermission(cameraActivity, android.Manifest.permission.CAMERA) !=
                PackageManager.PERMISSION_GRANTED && (ActivityCompat.checkSelfPermission(cameraActivity,
                        android.Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED)){
            ActivityCompat.requestPermissions(HomeActivity(), arrayOf(
                    android.Manifest.permission.CAMERA,
                    android.Manifest.permission.WRITE_EXTERNAL_STORAGE
            ), REQUESTCODECAMERA)
            return
        }
        cameraManager.openCamera(cameraId, stateCallback, null)
    }

    private fun takePicture() {
        if(cameraDevice == null)
            return
        val cameraManager: CameraManager = activity?.getSystemService(Context.CAMERA_SERVICE) as CameraManager
        val characteristics = cameraManager.getCameraCharacteristics(cameraDevice?.id)
        val jpegSizes : Array<Size>? =
            characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP).getOutputSizes(ImageFormat.JPEG)

        var width = 640
        var height = 480
        if(jpegSizes != null && jpegSizes.size > 0){
            width = jpegSizes[0].width
            height = jpegSizes[0].height
        }

        val imageReader: ImageReader = ImageReader.newInstance(width, height, ImageFormat.JPEG, 1)
        val outputSurface = ArrayList<Surface>(2)
        outputSurface.add(imageReader.surface)
        outputSurface.add(Surface(textureViewCamera.surfaceTexture))
        val captureBuilder : CaptureRequest.Builder? = cameraDevice?.createCaptureRequest(CameraDevice.TEMPLATE_STILL_CAPTURE)
        captureBuilder?.addTarget(imageReader.surface)
        captureBuilder?.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO)

        val rotation = activity?.windowManager?.defaultDisplay?.rotation
        captureBuilder?.set(CaptureRequest.JPEG_ORIENTATION, rotation?.let { ORIENTATIONS.get(it) })
        val tsLong : Long = System.currentTimeMillis()/ 1000
        val ts : String = tsLong.toString()

        file = File("${Environment.getExternalStorageDirectory()}/$ts.jpg")

        val readerListener : ImageReader.OnImageAvailableListener = ImageReader.OnImageAvailableListener {
            var image: Image = imageReader.acquireLatestImage()
            val byteBuffer: ByteBuffer = image.planes[0].buffer
            val bytes  = byteArrayOfInts(byteBuffer.capacity())
            byteBuffer.get(bytes)
            try {
                save(bytes)
            }catch (e: IOException){
                e.printStackTrace()
            }finally {
                image?.let { it.close() }
            }


        }
        imageReader.setOnImageAvailableListener(readerListener, backgroundHandler)

        val captureListener : CameraCaptureSession.CaptureCallback = object : CameraCaptureSession.CaptureCallback() {
            override fun onCaptureCompleted(session: CameraCaptureSession, request: CaptureRequest, result: TotalCaptureResult) {
                super.onCaptureCompleted(session, request, result)
                Toast.makeText(activity, "Saved", Toast.LENGTH_LONG).show()
                try{
                    createCameraPreview()
                }catch (e:CameraAccessException){
                    e.printStackTrace()
                }
            }
        }
        cameraDevice?.createCaptureSession(outputSurface, object: CameraCaptureSession.StateCallback(){
            override fun onConfigureFailed(session: CameraCaptureSession) {
                try {
                    session.capture(captureRequestBuilder.build(), captureListener, backgroundHandler)

                }catch (e: CameraAccessException){
                    e.printStackTrace()
                }
            }

            override fun onConfigured(session: CameraCaptureSession) {

            }

        }, backgroundHandler)

    }

    private fun save(bytes: ByteArray) {
        val outputStream: OutputStream = FileOutputStream(file)
        outputStream.write(bytes)
        outputStream.close()
    }

    fun byteArrayOfInts(vararg ints: Int) = ByteArray(ints.size) { pos -> ints[pos].toByte() }

    override fun onResume() {
        super.onResume()
        startBackgroundThread()
        if(textureViewCamera.isAvailable){
            openCamera()
        }else {
            textureViewCamera.surfaceTextureListener = textureListener
        }

    }

    private fun startBackgroundThread() {
        backgroundThread = HandlerThread("Camera Background")
        backgroundThread?.start()
        backgroundHandler = Handler(backgroundThread?.looper)
    }

    override fun onPause() {
        stopBackgroundThread()
        super.onPause()
    }

    protected fun stopBackgroundThread() {
        if(backgroundThread != null){
            backgroundThread?.quitSafely()
            backgroundThread?.join()
        }
        backgroundThread = null
        backgroundHandler = null

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        if(requestCode == REQUESTCODECAMERA){
            if(grantResults[0] == PackageManager.PERMISSION_DENIED){
                Toast.makeText(activity, "Sorry, camera permission is necessary", Toast.LENGTH_LONG).show()
            }
        }
    }

}
